### ISSUE TITLE
conceal does not re-calculate width and wrap lines.

### ISSUE THREAD

**Issue by mattn** [vim/vim#9443]
### Steps to reproduce

`:source` this code.
```
set columns=80 wrap conceallevel=3
syntax match XXX /X\+/ conceal cchar= 
finish


XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXYYYY
```

### Expected behaviour

`YYYY` in a line but
![image](https://user-images.githubusercontent.com/10111/147829328-ed307aff-9430-4d1a-b1b7-b85901c56643.png)

### Version of Vim

8.2.3877

### Environment

Windows 11. gvim.exe

**Comment by mattn** [comment]
This issue makes breaking wrapped text in popup-window.

![image](https://user-images.githubusercontent.com/10111/147830104-35380b74-cc6a-4dca-8d0a-c8d299d907e3.png)

**Comment by brammool** [comment]
> ### Steps to reproduce
> 
> `:source` this code.
> ```
> set columns=80 wrap conceallevel=3
> syntax match XXX /X\+/ conceal cchar= 
> finish
> 
> 
> XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXYYYY
> ```
> 
> ### Expected behaviour
> 
> `YYYY` in a line but
> ![image](https://user-images.githubusercontent.com/10111/147829328-ed307aff-9430-4d1a-b1b7-b85901c56643.png)

This was a choice made at some point in the past.  It is already quite
difficult to calculate where a line would break, and thus how many screen
lines a line if the file occupies.  If we would also take concealing
into account it gets even more complicated.  That is why wrapping is
decided on without considering concealing.

Since this was decided computers have become faster, the extra work
might not be a big problem now.  It's still going to be complicated.

-- 
Keep America beautiful.  Swallow your beer cans.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
At least, strange background-color in popup seems to be a bug. Probably, screen_line in screen.c:774 seems to be wrong.

**Comment by mtwebb** [comment]
While outside my current ability to fix, I will happily do any grunt work of testing PRs and the like. The ability to factor conceal when determining wrapping would have a large impact on how I use vim with long form writing.

**Comment by kflu** [comment]
I'd like to add a visual example here if that helps:

![image](https://user-images.githubusercontent.com/1031978/130737173-42425271-be36-4eff-af0a-02a369d6819f.png)

![image](https://user-images.githubusercontent.com/1031978/130737193-92844abc-4372-4989-b941-7806a10ef3cc.png)

This is how emacs markdown-mode looks like (with the equivalence of `set wrap`)

![image](https://user-images.githubusercontent.com/1031978/130737795-44e91e80-cb15-4f5d-a253-911060172a38.png)

![image](https://user-images.githubusercontent.com/1031978/130737833-d20ed4ee-2eaf-4f04-844f-a0dc24b8cff2.png)

**Comment by scateu** [comment]
+1

**Comment by mattn** [comment]
Perhaps I was able to solve this problem with a small impact.
![screenshot](https://user-images.githubusercontent.com/10111/169089883-de5d860a-5c30-4886-92cf-1e13b22ed3d8.gif)

https://github.com/vim/vim/pull/10442

**Comment by mattn** [comment]
![screenshot](https://user-images.githubusercontent.com/10111/169093793-79282e76-4432-48dd-ac37-1237cabea6c9.gif)

**Comment by pappasam** [comment]
@mattn if you have time, would you please test this PR? https://github.com/vim/vim/pull/20472

**Comment by pappasam** [comment]
related issue: https://github.com/neovim/neovim/issues/14409 (commenting just so they're linked, conversation across repos is hard to follow)
